### PR TITLE
Fixes #2382 Refresh Server Information in JoinGameScreen

### LINKED ISSUE #2382: Refresh button / functionality for server info on join server screen
At present the server info is only queried once then left in place on the join server page. Even if the server goes down or maybe is restarted with different modules the text remains the same. You can actually join the server fine even if you got info when it was down, clicking to join will go check again and all will be well.

I would like to have a button and/or functionality for when you would like refreshed information on a server. A refresh button by the server info would be good, alternatively or additionally the server info could automatically reset when you navigate away from the screen or view a different server.

Related issue (same area): #2377

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/35540620-refresh-button-functionality-for-server-info-on-join-server-screen?utm_campaign=plugin&utm_content=tracker%2F5279&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F5279&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by nihal111**
### Contains
Fix for #2382

### How to test

- Start a headless run along with a normal run
- normal run instance should show GREEN information for Worlds and Modules labels in localhost under Custom tab
- kill the headless run
- Worlds and Modules information should still be GREEN, even after host is killed.
- Hit R key
- This should "unselect" the selected option on the left panel and clicking on localhost now would show RED "Connection Failed!" as expected.

The above should work for Listed/Online servers as well. ("Connection Failed!" goes back to "Requested." once the R key is pressed).

Pressing the R key unselects the selected option from `visibleList` and further clears `extInfo` which contains the hash table (between items on the left panel and server information) that needs to be refreshed.

### Outstanding before merging

- [x] Need to add a refresh button somewhere.
- [x] Probably need to refresh list on every new click in the left panel. (done in latest commit)
- [ ] Need to add translation for word "Refresh" in all other languages.

### CODE REVIEW

**Comment by nihal111** [comment]
@Cervator @rzats Have a look. I need some help on creating that refresh button, though.

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by Cervator** [comment]
Works! Nice :-)

Just need to make it a button and automatically reselect the active server :-)

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by nihal111** [comment]
@rzats @Cervator Added the Refresh Button alongside the Join Button. Looks good?
The 'R' key is still functional. Should I remove it?
I have added the entry `"join-server-refresh": "Refresh"` to all language files. Need to update that. Should I look up translations and do that?

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/JoinGameScreen.java [hunk 1]
```diff
@@ -387,4 +407,19 @@ private String getModulesText(Future<ServerInfoMessage> info) {
...
 
+    public boolean onKeyEvent(NUIKeyEvent event) {
+        if (event.isDown() && event.getKey() == Keyboard.Key.R) {
+            refresh();
+            }
+            return false;
+        }
+
+    public void refresh() {
+        logger.info("Refresh");
```
Comment (rzats) [inline engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/JoinGameScreen.java]: I'd either remove this message or expand on it a bit (i.e. "Refreshing info for server X").

File: engine/src/main/resources/assets/ui/menu/joinGameScreen.ui [hunk 2]
```diff
@@ -1,338 +1,374 @@
 {
-    "type" : "joinGameScreen",
```
Comment (rzats) [inline engine/src/main/resources/assets/ui/menu/joinGameScreen.ui]: Back to four spaces, please! 😛 I'd also recommend setting your Eclipse/IDEA codestyle template to Terasology's (located at `config/eclipse/formatting.xml`).

File: engine/src/main/resources/assets/i18n/menu_cs.lang [hunk 3]
```diff
@@ -86,6 +86,7 @@
     "invert-y": "Invert Y Axis",
     "invert-z": "Invert Z Axis",
     "join-game-online": "Připojit se ke hře",
+    "join-server-refresh": "Refresh",
```
Comment (rzats) [inline engine/src/main/resources/assets/i18n/menu_cs.lang]: "join-server-refresh" should be placed just after "join-server-action" - this is the case in some of your files but not others.

File: engine/src/main/resources/assets/ui/menu/joinGameScreen.ui [hunk 4]
```diff
@@ -1,338 +1,374 @@
...
+                  },
+                  "position-bottom": {
+                    "target": "TOP",
+                    "widget": "buttonRow",
+                    "offset": 8
+                  }
+                }
+              },
+              {
+                "type": "UIButton",
```
Comment (rzats) [inline engine/src/main/resources/assets/ui/menu/joinGameScreen.ui]: Please delete this button, as it's a duplicate squeezed in between the other two.

image

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by Cervator** [comment]
Tested out again, works well, merged! Thanks :-)

**Pull request merged**